### PR TITLE
planner: Optimized EXTRACT(<UNIT> FROM <partition_column>) for partition pruning.

### LINKED ISSUE #54210: EXTRACT(YEAR FROM col) partitioning expression does not prune partitions
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```mysql
create table t (a datetime, b varchar(255)) partition by range (extract(year from a)) (partition p2020 values less than (2021), partition p2021 values less than (2022), partition pMax values less than (maxvalue));
insert into t values ('2020-12-31 23:59:59','2020-12-31 23:59:59'),('2021-01-01', '2021-01-01');
explain select * from t where a = '2020-12-31 23:59:59';
```

### 2. What did you expect to see? (Required)
That only partition p2020 would be used.
```mysql
tidb> explain select * from t where a = '2020-12-31 23:59:59';
+-------------------------+----------+-----------+-----------------+------------------------------------------+
| id                      | estRows  | task      | access object   | operator info                            |
+-------------------------+----------+-----------+-----------------+------------------------------------------+
| TableReader_7           | 10.00    | root      | partition:p2020 | data:Selection_6                         |
| └─Selection_6       | 10.00    | cop[tikv] |                 | eq(test.t.a, 2020-12-31 23:59:59.000000) |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t         | keep order:false, stats:pseudo           |
+-------------------------+----------+-----------+-----------------+------------------------------------------+

```
### 3. What did you see instead (Required)
All partitions where used.
```mysql
tidb> explain select * from t where a = '2020-12-31 23:59:59';
+-------------------------+----------+-----------+---------------+------------------------------------------+
| id                      | estRows  | task      | access object | operator info                            |
+-------------------------+----------+-----------+---------------+------------------------------------------+
| TableReader_7           | 10.00    | root      | partition:all | data:Selection_6                         |
| └─Selection_6       | 10.00    | cop[tikv] |               | eq(test.t.a, 2020-12-31 23:59:59.000000) |
|   └─TableFullScan_5 | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo           |
+-------------------------+----------+-----------+---------------+------------------------------------------+
```
### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v8.1.0
Edition: Community
Git Commit Hash: 945d07c5d5c7a1ae212f6013adfb187f2de24b23
Git Branch: HEAD
UTC Build Time: 2024-05-21 03:52:40
GoVersion: go1.21.10
Race Enabled: false
Check Table Before Drop: false
Store: tikv
```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #54210

Problem Summary:
the expression `EXTRACT(<time unit> FROM col)` is not handled for partition pruning.

### What changed and how does it work?
Treat the expression as prunable for all time units except WEEK (due to `default_week_format`).
Meaning all prunable expressions will work with EQ/`col = <date/datetime/time>`.
For DATE/DATETIME columns, it also optimizes YEAR and YEAR_MONTH to also work with ranges (<, <=, >, >=, BETWEEN), and for TIME columns it also optimizes HOUR, HOUR_MINUTE, HOUR_SECOND and HOUR_MICROSECOND for ranges.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/rule_partition_processor.go [hunk 1]
```diff
@@ -1553,6 +1608,16 @@ func replaceColumnWithConst(partFn *expression.ScalarFunction, con *expression.C
 			return partFn
 		}
 	}
+	if partFn.FuncName.L == ast.Extract {
```
[the remark this answers is not in the document]
Comment (Defined2014) [reply pkg/planner/core/rule_partition_processor.go:1611]: Or change it to `switch ... case ...`

File: pkg/planner/core/casetest/partition/partition_pruner_test.go [hunk 2]
```diff
@@ -233,3 +234,317 @@ func TestPointGetIntHandleNotFirst(t *testing.T) {
 	tk.MustQuery("select * from t WHERE a BETWEEN 13 AND 13").Check(testkit.Rows("1 13 1"))
 	tk.MustQuery(`select * from t`).Check(testkit.Rows("1 13 1"))
 }
+
+func TestPruneExtractYear(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mjonss) [reply pkg/planner/core/casetest/partition/partition_pruner_test.go]: Right, I created it first and then it is covered again, with the more extensive test cases. I have removed this test.

**Review by Defined2014 — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**